pkg/sdn/registry/netnamespace/migration.go

```diff
@@ -0,0 +1,105 @@
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/storage"
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+
+	"github.com/openshift/openshift-sdn/pkg/netid"
+	"github.com/openshift/origin/pkg/sdn/api"
+)
+
+const (
+	etcdNetNamespacePrefix string = "/registry/sdnnetnamespaces"
+)
+
+type Migrate struct {
+	etcdHelper storage.Interface
+	client     kclient.NamespaceInterface
+}
+
+func NewMigrate(etcdHelper storage.Interface, client kclient.NamespaceInterface) *Migrate {
+	return &Migrate{
+		etcdHelper: etcdHelper,
+		client:     client,
+	}
+}
+
+func (m *Migrate) Run() error {
+	var netnsList api.NetNamespaceList
+
+	err := m.etcdHelper.List(context.TODO(), etcdNetNamespacePrefix, "", storage.Everything, &netnsList)
+	if err != nil {
+		return fmt.Errorf("List network namespaces failed: %v, please retry the operation", err)
+	}
+
+	errList := []error{}
+	for _, netns := range netnsList.Items {
+		if err := m.migrateNetNamespace(&netns); err != nil {
+			errList = append(errList, err)
+			continue
+		}
+	}
+
+	if len(errList) > 0 {
+		errList = append(errList, fmt.Errorf("Unexpected errors during NetNamespace migration, please retry the operation"))
+	}
+	return kerrors.NewAggregate(errList)
+}
+
+func (m *Migrate) deleteNetNamespace(netns *api.NetNamespace) {
+	var out api.NetNamespace
+	if err := m.etcdHelper.Delete(context.TODO(), path.Join(etcdNetNamespacePrefix, netns.ObjectMeta.Name), &out, nil); err != nil {
+		glog.Errorf("Failed to delete NetNamespace: %v, error: %v", netns, err)
+	}
+}
+
+func (m *Migrate) migrateNetNamespace(netns *api.NetNamespace) error {
+	// Get corresponding namespace for the NetNamespace object
+	ns, err := m.client.Get(netns.NetName)
+	// If corresponding namespace no longer exists, nothing to migrate in this case.
+	if errors.IsNotFound(err) {
+		m.deleteNetNamespace(netns)
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Get namespace %q failed: %v", netns.NetName, err)
+	}
+
+	// Set netid as namespace annotation
+	if err := netid.SetVNID(ns, netns.NetID); err != nil {
+		return fmt.Errorf("Set netid %d for namespace %q failed: %v", netns.NetID, ns.Name, err)
+	}
+
+	// Update namespace
+	updatedNs, err := m.client.Update(ns)
+	if errors.IsNotFound(err) {
+		m.deleteNetNamespace(netns)
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Update namespace %q failed: %v", ns.Name, err)
+	}
+
+	// Validate
+	if id, err := netid.GetVNID(updatedNs); err == nil {
+		if id == netns.NetID {
+			glog.Infof("Migrated netid %d from NetNamespace to annotation on namespace %q", netns.NetID, ns.Name)
+		} else {
+			return fmt.Errorf("Failed to migrate netid from NetNamespace to namespace %q, expected netid %d but got %d", ns.Name, netns.NetID, id)
+		}
+	} else {
+		return fmt.Errorf("Failed to migrate netid %d from NetNamespace to annotation on namespace %q", netns.NetID, ns.Name)
+	}
+
+	// Delete processed NetNamespace object
+	m.deleteNetNamespace(netns)
```

Comment:
Discussed with @liggitt on IRC, we need some upgrade order for masters/nodes that allows no downtime. 
If we upgrade master first, deleting NetNamespaces as part of migration will affect old nodes until they are upgraded (this could be few mins to hours or even more) and this is not acceptable. If we upgrade nodes first, we will miss NetNamespace events from old master until it is upgraded.
I didn't like the idea of GET NetNamespace to do lookup on namespace for vnid annotation and falling back to old persisted NetNamespace info in etcd if vnid annotation doesn't exist. Worried that if vnid annotation is missing for some reason, we may return stale/incorrect vnid from etcd.
I'm thinking of taking this approach:
- Make new nodes to watch for NetNamespaces along with Namespaces but react to NetNamespace events only if there is no openshift.io/netns.migrated=true annotation.
- Master as part of migration will add openshift.io/netns.migrated=true annotation to NetNamespace before deletion.
With these above changes, migration steps could be:
- Upgrade all nodes first (in any order)
- Upgrade all masters (sdn controller runs on one of the master anyway)

do you see any issues with this approach?